// tauri-plugin-posthog-api/dist-js/index.js
import { invoke } from '@tauri-apps/api/core';
import posthog from 'posthog-js';

/**
 * PostHog client for Tauri applications
 * Wraps PostHog JS SDK with automatic configuration from Tauri backend
 */
class PostHogTauri {
    /**
     * Initialize PostHog with configuration from Tauri backend
     * This is called automatically on first use
     */
    static async init() {
        if (this.initialized)
            return;
        if (this.initPromise)
            return this.initPromise;
        this.initPromise = this._performInit();
        return this.initPromise;
    }
    static async _performInit() {
        try {
            // Get configuration from Tauri backend
            const config = await invoke('plugin:posthog|get_config');
            // Initialize PostHog JS SDK with backend configuration
            posthog.init(config.apiKey, {
                api_host: config.apiHost,
                ...config.options
            });
            // Sync any existing distinct_id from Rust backend
            const existingDistinctId = await invoke('plugin:posthog|get_distinct_id');
            if (existingDistinctId) {
                // Apply the existing distinct_id to PostHog JS SDK without triggering backend call
                posthog.identify(existingDistinctId);
            }
            this.initialized = true;
        }
        catch (error) {
            console.error('Failed to initialize PostHog:', error);
            throw new Error(`PostHog initialization failed: ${error}`);
        }
    }
    /**
     * Capture an event with optional properties
     * @param event - The event name
     * @param properties - Event properties (optional)
     */
    static async capture(event, properties) {
        await this.init();
        posthog.capture(event, properties);
    }
    /**
     * Identify a user with a distinct ID and optional properties
     * This method synchronizes the distinct_id between Rust backend and PostHog JS SDK
     * @param distinctId - The unique identifier for the user
     * @param properties - User properties (optional)
     */
    static async identify(distinctId, properties) {
        await this.init();
        // First update the Rust backend state
        await invoke('plugin:posthog|identify', {
            distinct_id: distinctId,
            properties: properties
        });
        // Then update PostHog JS SDK (but skip properties since they're handled by backend)
        posthog.identify(distinctId);
    }
    /**
     * Create an alias for the current user
     * @param alias - The alias to create
     */
    static async alias(alias) {
        await this.init();
        posthog.alias(alias);
    }
    /**
     * Reset the current user (clears distinct ID and other user data)
     * This method synchronizes the reset between Rust backend and PostHog JS SDK
     */
    static async reset() {
        await this.init();
        // First reset the Rust backend state
        await invoke('plugin:posthog|reset');
        // Then reset PostHog JS SDK
        posthog.reset();
    }
    /**
     * Get the current distinct ID
     * Returns the distinct_id from Rust backend as the source of truth
     */
    static async getDistinctId() {
        await this.init();
        const backendDistinctId = await invoke('plugin:posthog|get_distinct_id');
        return backendDistinctId || posthog.get_distinct_id();
    }
    /**
     * Group identify - associate user with a group
     * @param groupType - The type of group (e.g., 'company', 'project')
     * @param groupKey - The key for the group
     * @param properties - Group properties (optional)
     */
    static async groupIdentify(groupType, groupKey, properties) {
        await this.init();
        posthog.group(groupType, groupKey, properties);
    }
    /**
     * Set person properties
     * @param properties - Properties to set on the person
     */
    static async setPersonProperties(properties) {
        await this.init();
        posthog.setPersonProperties(properties);
    }
    /**
     * Feature flag methods
     */
    static async isFeatureEnabled(flagKey) {
        await this.init();
        return posthog.isFeatureEnabled(flagKey) || false;
    }
    static async getFeatureFlag(flagKey) {
        await this.init();
        return posthog.getFeatureFlag(flagKey);
    }
    static async getFeatureFlagPayload(flagKey) {
        await this.init();
        return posthog.getFeatureFlagPayload(flagKey);
    }
    static async reloadFeatureFlags() {
        await this.init();
        posthog.reloadFeatureFlags();
    }
    /**
     * Page view tracking
     * @param properties - Page properties (optional)
     */
    static async capturePageView(properties) {
        await this.init();
        posthog.capture('$pageview', properties);
    }
    /**
     * Get the PostHog JS SDK instance (advanced usage)
     * Ensures PostHog is initialized before returning
     */
    static async getInstance() {
        await this.init();
        return posthog;
    }
    /**
     * Check if PostHog has been initialized
     */
    static isInitialized() {
        return this.initialized;
    }
    /**
     * Manually initialize PostHog (optional - will be called automatically)
     */
    static async initialize() {
        return this.init();
    }
}
PostHogTauri.initialized = false;
PostHogTauri.initPromise = null;

export { PostHogTauri as PostHog, PostHogTauri, PostHogTauri as default };
